crates/dbsp/src/trace/spine_async/merger.rs

```diff
@@ -0,0 +1,150 @@
+//! A compactor thread that merges the batches for the spine-fueled trace.
+
+use crate::storage::backend::metrics::{COMPACTION_DURATION, COMPACTION_SIZE, TOTAL_COMPACTIONS};
+use crate::storage::backend::StorageError as Error;
+use crate::trace::spine_async::BatchIdent;
+use crate::trace::{Batch, Filter, Merger};
+use crate::Runtime;
+use crossbeam::channel::{bounded, unbounded, Receiver, RecvError, Sender};
+use metrics::{counter, histogram};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::Instant;
+
+pub(super) enum MergeOperation<B>
+where
+    B: Batch,
+{
+    Merge(
+        BatchIdent,
+        Vec<Arc<B>>,
+        Option<Filter<B::Key>>,
+        Option<Filter<B::Val>>,
+    ),
+}
+
+pub(super) enum MergeResult<B>
+where
+    B: Batch,
+{
+    MergeCompleted(Result<(BatchIdent, B), Error>),
+}
+
+pub(crate) struct BatchMerger<B>
+where
+    B: Batch,
+{
+    /// A handle to initiate merge operations on the merger thread.
+    sender: Sender<MergeOperation<B>>,
+    /// A handle to receive completed merge operations from the merger thread.
+    receiver: Receiver<MergeResult<B>>,
+    /// In progress merges.
+    _in_progress: Vec<(MergeOperation<B>, <B as Batch>::Merger)>,
+    /// To make sure we use the same batch type as the spine using this.
+    _phantom: PhantomData<B>,
+}
+
+impl<B> BatchMerger<B>
+where
+    B: Batch,
+{
+    /// We sample the queue length only once every `QUEUE_LENGTH_SAMPLE_RATE` merges.
+    const QUEUE_LENGTH_SAMPLE_RATE: usize = 1_000;
+
+    /// Size of the incoming merge queue.
+    const RX_QUEUE_SIZE: usize = 64;
+    /// How many concurrent merges we allow.
+    const CONCURRENT_MERGES: usize = 1;
+
+    pub(crate) fn new() -> Self {
+        let (new_merge_sender, new_merge_receiver) = bounded(Self::RX_QUEUE_SIZE);
+        let (completed_merge_sender, completed_merge_receiver) = unbounded();
+        let mut merge_cnt = 0;
+
+        Runtime::spawn_compactor(move || loop {
+            let op: Result<MergeOperation<B>, RecvError> = new_merge_receiver.recv();
+            merge_cnt += 1;
+            if merge_cnt % Self::QUEUE_LENGTH_SAMPLE_RATE == 0 {
+                histogram!(COMPACTION_DURATION).record(new_merge_receiver.len() as f64);
+            }
+
+            match op {
+                Ok(MergeOperation::Merge(key, mut batches, key_filter, value_filter)) => {
+                    // TODO: currently batches.len() is always 2, except when we call consolidate.
+                    // In the future we should make add support for multi-way merges so we don't
+                    // have to merge so inefficiently in presence of many batches.
```

Comment:
If we have 2-way merges but multiple batches it's more efficient to always merge the two shortest batches, like how the `merge_batches` function does.

Reply:
Oh but hmm we're always working with batches that are approximately the same size? Except that if we're doing multiple merges then the earlier ones might discard lots of records or even add to an empty batch... Hmm...

Reply:
I'm not sure I follow the thought; but yes to "we're always working with batches that are approximately the same size" and for "or even add to an empty batch" the spine code should discards attempts at inserting empty batches

Reply:
I mean that if a and b are nonempty, then a merged with b might be empty (or, more likely, just a lot smaller than the sum of their individual lengths).

Reply:
right, do you want me to sort them by length before starting the merges? 

fwiw it didn't bother with it because currently len()=2 but it's an easy addition

Reply:
Let's leave it until n>2 is possible.

Reply:
(I was thinking out loud, I guess, sorry)